### PR TITLE
fix: Lazy-load grant & grantRound metadata

### PR DESCRIPTION
**Pull request by gdixon**
This PR:

- ensures that we are lazy-loading metadata and shows placeholder content while we're waiting for the metaPtr content to be fetched from ipfs (Closes: #510)
- reduces the number of RPC calls we make on init
- allows for `metaPtr.protocol === 0` to default to placeholder images (Fixes: #532 && Fixes: #537)
- fixes typings on `cachedWhiteList` (`any` -> `LocalForageData`)
- only `checkIsDeployed` if were on dev (Fixes: #536)

--

### CODE REVIEW

**Review by thelostone-mc — APPROVED** [review]

File: app/src/utils/utils.ts [hunk 1]
```diff
@@ -85,7 +85,7 @@ export function cleanTwitterUrl(url: string | undefined) {
 }
 
 // Given a metadata pointer object, return a stringified version that can be used as an object key
-export function metadataId(metaPtr: MetaPtr): string {
+export function metadataId(metaPtr: MetaPtr = { protocol: 0, pointer: '' }): string {
```
Comment (mds1) [inline app/src/utils/utils.ts:88]: What's the use case for setting this default, i.e. when would we call `metadataId()` to return a value of `0-`, without providing an input?

**Review by apbendi — COMMENTED** [review]
@gdixon thanks for tackling this! Testing out locally, I'm not seeing the intended effect play out practically. The page doesn't seem to load until all content is available, and in the network console, I see all the IPFS assets have fetched before the loading spinner disappears and the UI becomes available.

lazy-load

**Comment by gdixon** [comment]
> @gdixon thanks for tackling this! Testing out locally, I'm not seeing the intended effect play out practically. The page doesn't seem to load until all content is available, and in the network console, I see all the IPFS assets have fetched before the loading spinner disappears and the UI becomes available.
> 
> lazy-load

Could you try again with the changes I've just pushed? I had set a timeout around the `fetchMetaPtrs` during testing, and everything seemed okay, but you are right! It was till waiting on the resolution - the changes in f8dc6db should fix that!

File: app/src/utils/utils.ts (hunk 1 above)
Comment (gdixon) [reply app/src/utils/utils.ts:88]: This handles the case where metadata hasn't yet loaded, specifically on the cart.

File: app/src/store/data.ts [hunk 2]
```diff
@@ -209,73 +243,76 @@ export default function useDataStore() {
...
 
-    // Set up watchers on the grantRounds
-    grantRoundsList.forEach(async (grantRound) => {
-      // open the rounds contract
-      const roundContract = new Contract(grantRound.address, GRANT_ROUND_ABI, DEFAULT_PROVIDER);
-      // open the rounds contract
-      const matchingTokenContract = new Contract(await roundContract.matchingToken(), ERC20_ABI, DEFAULT_PROVIDER);
-      // attach listeners to the rounds
+    // init listeners but dont block execution
+    setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (gdixon) [reply app/src/store/data.ts:276]: Ahh, good catch, I've moved this lower (to the map on the second batch of listeners).

File: app/src/components/GrantCard.vue [hunk 3]
```diff
@@ -17,7 +17,7 @@
       <div class="aspect-w-16 aspect-h-9 shadow-light">
         <img
           class="w-full h-full object-center object-cover group-hover:opacity-90"
-          :src="ptrToURI(logoPtr) || '/placeholder_round.svg'"
+          :src="(logoPtr?.protocol !== 0 && ptrToURI(logoPtr)) || '/placeholder_grant.svg'"
```
[the remark this answers is not in the document]
Comment (gdixon) [reply app/src/components/GrantCard.vue:20]: Awesome! Yeah I will make those changes asap!

**Review by apbendi — APPROVED** [review]
Looks good and works well. Nice job @gdixon.

File: app/src/components/GrantCard.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (gdixon) [reply app/src/components/GrantCard.vue:20]: I've created an issue to track this: https://github.com/dcgtc/dgrants/issues/541

**Review by mds1 — APPROVED** [review]

**Pull request merged**